pkg/controller/cronjob/utils_test.go

```diff
@@ -277,84 +277,182 @@ func TestByJobStartTime(t *testing.T) {
+			expectedRecentTime:     deltaTimeAfterTopOfTheHour(300 * time.Minute),
 			expectedNumberOfMisses: 5,
 		},
 		{
 			name: "rogue cronjob",
-			args: args{
-				earliestTime: deltaTimeAfterTopOfTheHour(time.Second * 10),
-				now:          *deltaTimeAfterTopOfTheHour(time.Hour * 1000000),
-				schedule:     "59 23 31 2 *",
+			cj: &batchv1.CronJob{
+				ObjectMeta: metav1.ObjectMeta{
+					CreationTimestamp: metav1.NewTime(*deltaTimeAfterTopOfTheHour(10 * time.Second)),
+				},
+				Spec: batchv1.CronJobSpec{
+					Schedule: "59 23 31 2 *",
+				},
 			},
-			expectedTime:           nil,
+			now:                    *deltaTimeAfterTopOfTheHour(1000000 * time.Hour),
+			expectedRecentTime:     nil,
 			expectedNumberOfMisses: 0,
 			wantErr:                true,
 		},
+		{
+			name: "earliestTime being CreationTimestamp and LastScheduleTime",
+			cj: &batchv1.CronJob{
+				ObjectMeta: metav1.ObjectMeta{
+					CreationTimestamp: metav1TopOfTheHour,
+				},
+				Spec: batchv1.CronJobSpec{
+					Schedule: "0 * * * *",
+				},
+				Status: batchv1.CronJobStatus{
+					LastScheduleTime: &metav1TopOfTheHour,
+				},
+			},
+			now:                  *deltaTimeAfterTopOfTheHour(30 * time.Second),
+			expectedEarliestTime: topOfTheHour(),
+			expectedRecentTime:   nil,
+		},
+		{
+			name: "earliestTime being LastScheduleTime",
+			cj: &batchv1.CronJob{
+				ObjectMeta: metav1.ObjectMeta{
+					CreationTimestamp: metav1TopOfTheHour,
+				},
+				Spec: batchv1.CronJobSpec{
+					Schedule: "*/5 * * * *",
+				},
+				Status: batchv1.CronJobStatus{
+					LastScheduleTime: &metav1HalfPastTheHour,
+				},
+			},
+			now:                  *deltaTimeAfterTopOfTheHour(31 * time.Minute),
+			expectedEarliestTime: deltaTimeAfterTopOfTheHour(30 * time.Minute),
+			expectedRecentTime:   nil,
+		},
+		{
+			name: "earliestTime being LastScheduleTime (within StartingDeadlineSeconds)",
+			cj: &batchv1.CronJob{
+				ObjectMeta: metav1.ObjectMeta{
+					CreationTimestamp: metav1TopOfTheHour,
+				},
+				Spec: batchv1.CronJobSpec{
+					Schedule:                "*/5 * * * *",
+					StartingDeadlineSeconds: &oneMinute,
+				},
+				Status: batchv1.CronJobStatus{
+					LastScheduleTime: &metav1HalfPastTheHour,
+				},
+			},
+			now:                  *deltaTimeAfterTopOfTheHour(31 * time.Minute),
+			expectedEarliestTime: deltaTimeAfterTopOfTheHour(30 * time.Minute),
+			expectedRecentTime:   nil,
+		},
+		{
+			name: "earliestTime being LastScheduleTime (outside StartingDeadlineSeconds)",
+			cj: &batchv1.CronJob{
+				ObjectMeta: metav1.ObjectMeta{
+					CreationTimestamp: metav1TopOfTheHour,
+				},
+				Spec: batchv1.CronJobSpec{
+					Schedule:                "*/5 * * * *",
+					StartingDeadlineSeconds: &oneMinute,
+				},
+				Status: batchv1.CronJobStatus{
+					LastScheduleTime: &metav1HalfPastTheHour,
+				},
+			},
+			includeSDS:           true,
+			now:                  *deltaTimeAfterTopOfTheHour(32 * time.Minute),
+			expectedEarliestTime: nil,
```

Comment:
missing `expectedEarliestTime`

Reply:
from the code it seems it will be set to schedulingDeadline. Although it has not run in that time. Is there a reason to discard/not to test this value?